### PR TITLE
Add support for software Pulse-width Modulation (PwM) on all channels and much, much more

### LINKED ISSUE #3: Pulse width Modulation (PwM)
This will probably require modifications to the kernel

### LINKED ISSUE #4: There are not many examples and some are out of date

### LINKED ISSUE #13: Move TODOs out of RPi/_GPIO.py comments

### LINKED ISSUE #17: mechanism to allow non-root access to gpio pins
We may want to add a mechanism to enable/disable non-root access to the gpio pins. I'm not sure if this should be the default behavior or not, but perhaps we could include a simple toggling utility depending how how we implement this

### LINKED ISSUE #22: spec and SPECS sound too similar
I think it makes more sense to change the latter given that is used for particular packaging case whereas the functional and technical spec is general to the project

### LINKED ISSUE #24: A few constants are missing
Specifically, we want I2C, SPI, HARD_PWM, SERIAL, RPI_INFO, and RPI_REVISION accessible under RPi.GPIO

### PR DESCRIPTION
**Pull request by theyoyojo**
```
first major patch:
	pwm:
	- implement PwM interface as per spec
	- implement basic software PwM functionality
	- add logic tests for PwM interface
	- extend threading object hierarchy to integrate PwM thread
		control with existing thread control flow logic
		this includes remaming _PollThread as _LineThread
	- add _line_thread_* thread type constants
	- add thread_{stop,start} interface to _Line to avoid direct
		manipulation of _LineThread
	- add dutycycle and frequency fields to class _Line
	- update spec to reflect new internal changes (in line_* interface)
	- bump spec revision number to 2
	- s/line_start_poll/line_poll_start/g
	- add dictionary mapping thread types to thread entry points

	examples:
	- fix broken examples
	- add flashing light example
	- deprecate morese code example
	- remove old wrap.py prototype
	- fix and enforce style in examples
	- set executable bit on all examples

	misc:
	- add a note about notes.sh to README
	- add header to RPi/_GPIO
	- mention that the project is free software in the README and header
	- export the PWM name in RPi/GPIO/__init__.py for access via GPIO name
	- s/is_all_bools/is_all_bools_or_directions/g
		this function now returns True when GPIO.HIGH or GPIO.LOW
		are the data or in the list
	- fix GPIO.HIGH and GPIO.LOW all being interpreted as True by output()
	- remove the directory named 'misc' that contained files
		to reproduce a random bug and should not have been
		in the master branch
	- fix test-style.sh not scanning tests and examples
	- fix a lot of style errors in the tests
	- replace some magic numbers with hopefully more descriptive constants
	- remove large ascii art TODO and redundant comments at the top of _GPIO.py
 	- add script to enable non-root access to gpio pins
	- add missing constants (I2C, SERIAL, HARD_PWM, SPI)
	- `mv SPECS packaging` to delete name similarity with spec directory

first minor patch:
	rename SPECS directory to packaging to fix (#22) as this directory is for
	files that pertain to distro-specific packaging

second minor patch:
	add script to enable non-root access to gpio pins

third minor patch:
	add missing constants I2C, SPI, HARD_PWM, SERIAL

second major patch:
	.coveragerc:
	- report coverage information on RPi/core.py instead
	  of RPi/_GPIO.py due to file name change

	Makefile:
	- %s/unit-cov/test-cov/g
	- removed an extraneous comment

	RPi/GPIO/__init__.py:
	RPi/GPIO_DEVEL/__init__.py:
	- Added header docstrings
	- formatted and sorted API imports

	RPi/_GPIO.py -> RPi/core.py:
	- File renamed to clarify purpose
	- convert header to docstring format
	- sort and organize functions and constants
	- fix documentation errors and gramatical mistakes
	- add automatic invokation of cleanup() at interpreter exit
	- add RPi.GPIO.{RPI_INFO, RPI_REVISION, VERSION} constants
	- fix various cases of mishandling of iterable channel and value
	  lists
	- fix pwm thread never releasing lock if thread dies due to
	  exception when the main thread revokes access, however I
	  suspect that this actually be a surface fix for another bug,
	  will look into further (#29)
	- small tweaks to improve modularity
	- add support for single value and iterable lists of values for
	  RPi.GPIO.cleanup()
	- allow PwM to successfully start without explicitly setting up
	  a channel as an output because this library has a more relaxed
	  attitude about those things
	- add warn on invalid call to PWM.start()
	- fix board numbering mode being actually pretty broken due to
	  inconsistent mishandling of channel inputs of various forms
	- fix line_get_value being actually very broken and not
	  retuning anything and therefore always returning None
	- fix gpio_function() being actually entirely broken so it
	  returns the current function representing what is going on on
	  a GPIO channel as RPi.GPIO would, but this requires further
	  testing (#30)
	- fix RPi.GPIO.PWM.start() not returning the truth of whether a
	  PwM thread has been successfully started or not

	examples/callback2.py:
	- add shabang

	examples/flash18.py:
	- remove debuginfo output toggle

	examples/input_examples.py:
	examples/output_examples.py:
	examples/pwm2.py:
	examples/pwm_usage.py:
	examples/random_usage.py:
	examples/try_import.py:
	- copy example code from Ben Croston's RPi.GPIO wiki on
	  sourceforge
	- see the files for the sources
	- ensure that the code works with this library. I found quite a
	  few bugs integrating these files
	- fix a few mistakes in the code
	- fix style

	examples/morse.py:
	- update morse code to current API version
	- add basic input validation
	- fix style

	examples/pwm.py:
	- renamed to pwm1.py for consistency

	examples/pwm1.py:
	- possible issue: takes two Ctl+C's to stop
	- this may be revelvant to (#29)

	examples/pwm3.py:
	examples/pwm4.py:
	- Copy example code from some forum
	- see the files for the sources

	non_root_permission.sh:
	- expand note on lack of persistence of effects post reboot

	spec/spec.tex:
	- fix some typos
	- tweak style and title
	- add entry for channel_valid_or_die
	- update some of the technical spec

	spec/sources.bib:
	- replace \url with \texttt

	spec/spec.pdf:
	- regenerate document with 1 inch margins

	test-style.sh:
	- scan RPi/core.py instead of RPi/_GPIO.py due to name change
	- add scan of examples/morese.py since it is again working

	tests/test_gpio.py:
	- style tweaks
	- fix test_gpio_function by making it actually test for the
	  feature that now is possibly implemented correctly
	- modify test_setdebug info to turn it off before the next call
	  to Reset to avoid printing a lot of text to the terminal
	- extend test_cleanup to validate the new features of cleanup()

	tests/test_pwm.py:
	- more thoroughly test PWM.start()
```
tl;dr: Anyway, here's software PwM and also pretty much the rest of the missing features

fixes #3
fixes #4
fixes #13
fixes #17
fixes #24 
fixes #22

### CODE REVIEW

File: RPi/core.py [hunk 1]
```diff
@@ -111,32 +135,58 @@ def bias_flag(const):
     _line_mode_as_is:   -662,
 }
 
+# line thread types, these threads can be run on an individual line
+_line_thread_none       = 0
+_line_thread_poll       = 1 << 1
+_line_thread_pwm        = 1 << 2
```
Comment (fabrizio8) [inline RPi/core.py:141]: How about:
```
_line_thread_poll       = 0b10
_line_thread_pwm        = 0b100
```
instead?

**Review by fabrizio8 — APPROVED** [review]

File: RPi/core.py (hunk 1 above)
Comment (theyoyojo) [reply RPi/core.py:141]: the only reason to use bits here is if we want to support running both poll and pwm threads and I don't know if that really makes sense, but that would be slightly faster since there are two less ops (in theory)

**Comment by theyoyojo** [comment]
yeet
